cpu/native/Makefile.dep

```diff
@@ -1,3 +1,15 @@
+ifeq (FreeBSD,$(OS))
+  ifeq ($(OS_ARCH),amd64)
+    DISABLE_LIBSTDCPP ?= 1
+  endif
+  ifeq ($(DISABLE_LIBSTDCPP),1)
+    # static C++ constructors need guards for thread safe initialization
+    ifneq (,$(filter cpp,$(FEATURES_USED)))
+      USEMODULE += cxx_ctor_guards
+    endif
+  endif
+endif
+
```

Comment:
```suggestion
ifeq (,$(filter libstdcpp,$(FEATURES_USED)))
  # static C++ constructors need guards for thread safe initialization
  ifneq (,$(filter cpp,$(FEATURES_USED)))
    USEMODULE += cxx_ctor_guards
  endif
endif
```

Reply:
Is this in all cases or only for FreeBSD? If its only for FreeBSD then the first ifdef should be kept as well.

Reply:
It's only for FreeBSD 64-bit, yes. Otherwise this could be moved to a more global `Makefile.dep`, I believe.

Reply:
And why so? Only arch using them so far is atmega? Why did you add it for FreeBSD?

Reply:
It provides the guards needed to call constructors during the initialization of statically allocated class instances. The C++ compiler will generate call to those utility functions. Usually those are implemented by the toolchain (e.g. as part of libstdc++). On FreeBSD the default libstdc++ is sadly not compatible with `-m32`, so undefined references to those implementation will occur while linking. The module is a compatible drop-in replacement for platforms without (working) libstdc++, such as ATmega or FreeBSD.

Reply:
But then this dependency could be handled globally with the proposal by @fjmolinas put e.g. to `Makefile.dep`, if I read what you wrote correctly. The requirement to include `cxx_ctor_guards` is solely dependent of the fact if there is `cpp` provided, but not `libstdcpp` and not on the underlying board/CPU directly (other than them being the ones to provide those features of course).

Reply:
In that case I would move this to `Makefile.dep` with a comment as @miri64 suggested.

Reply:
ESP32 has a platform specific implementation of the `cxa_ctor_guards` in `cpu/esp32/cxx`. But note that ESP32 does have a `libstdc++`, it only lacks the guards. So the logic `if (cpp && !libstdcpp)` would not work out. That said, I'm not against handling this at a global level. I'm just concerned that this might not be feasible.

Btw: We might want to drop `cpu/esp32/cxx` in favor of the platform independent `cxa_ctor_guards`. I think the ESP32 implementation is not correct. According to the documentation it is not thread safe, but thread safety was the very point those guards are there.

Reply:
To avoid confusion: I'm not against moving this to `Makefile.dep`. (I have no personal preferences on where to put this.)

Reply:
> Btw: We might want to drop `cpu/esp32/cxx` in favor of the platform independent `cxa_ctor_guards`. I think the ESP32 implementation is not correct. According to the documentation it is not thread safe, but thread safety was the very point those guards are there.

Summon @gschorcht?

Reply:
But true... with that knowledge having the condition `(cpp && !libstdcpp)` in `Makefile.dep` to include `cxx_ctor_guards` makes little sense :-/.

Reply:
Then I keep it here. If I remove the dependency on `DISABLE_LIBSTDCPP` I make dependent on @fjmolinas answer to https://github.com/RIOT-OS/RIOT/pull/14631#issuecomment-665123344

Reply:
> Btw: We might want to drop `cpu/esp32/cxx` in favor of the platform independent `cxa_ctor_guards`.

Indeed, it's just a leftover required by former ESP32 toolchain versions that used POSIX threads. Although `cxa_ctor_guards` were built-in, they had to be replaced since they used the `pthread_once` function for singleton objects initialization where the parameter `once` was of incompatible type with that provided by RIOT's `pthread` module.

We are now using a self-compiled ESP32 toolchain version that no longer uses POSIX threads for thread-safety. That is, the toolchain does no longer guarantee thread-safety, as is the case with other platforms.

> I think the ESP32 implementation is not correct. According to the documentation it is not thread safe, but thread safety was the very point those guards are there.

Aborting if a thread tries to access the variable during its initialization by another thread was the only way to get it to work at that time :worried:

Reply:
The `cxx_ctor_guards` just use an rmutex for locking. That way the same thread is able to call the ctor of classes that call the ctor of other classes within their ctor and is thread safe. If those guards are indeed still needed, the module `cxx_ctor_guards` should be a drop in replacement for every platform. (It only uses a single byte of whatever is passed as `__guard`, so no matter how small this type is defined on a specific platform, it will still work.)

Reply:
> The `cxx_ctor_guards` just use an rmutex for locking.

Ah, there is a new module `cxx_ctor_guards`. Sorry, I didn't notice that because I've been pretty busy the last few weeks rebasing and providing the new GPIO API after a half year break.

I don't remember exactly, but using a mutex for the ESP32 `cxx_ctor_guards` led to deadlocks. Anyway, we can drop it now. I will provide a PR.